src/TestCase.php

```diff
@@ -2,64 +2,93 @@
 
-    /** @var RequestFactory */
-    private $requestFactory;
+    /** @var MessageFactory */
+    private $messageFactory;
 
-    /** @var FutureHttpPool */
-    private $futureHttpPool;
+    /** @var Semaphore */
+    private $semaphore;
 
-    final public function __construct(RequestFactory $requestFactory, FutureHttpPool $pool)
-    {
-        $this->requestFactory = $requestFactory;
-        $this->futureHttpPool = $pool;
-    }
+    /** @var Client */
+    private $client;
 
-    /**
-     * @return FutureHttpPool
-     */
-    public function getFutureHttpPool()
+    final public function __construct(MessageFactory $messageFactory, Semaphore $semaphore)
     {
-        return $this->futureHttpPool;
+        $this->messageFactory = $messageFactory;
+        $this->semaphore = $semaphore;
     }
 
     /**
      * Run before each test.
      *
      * Allow to set default services and context, and also decorate the http async client.
      *
-     * @param HttpAsyncClient $asyncClient
+     * @param Client $asyncClient
      *
-     * @return HttpAsyncClient
+     * @return Client
      */
-    public function setUp(HttpAsyncClient $asyncClient)
+    public function setUp(Client $asyncClient): Client
     {
         return $asyncClient;
     }
 
+    final public function initialize()
+    {
+        $this->client = $this->setUp(new DefaultClient());
+    }
+
     /**
      * Allow to test a rejection or a resolution of an async call.
      *
-     * @param RequestInterface $requestInterface
+     * @param RequestInterface $request
      *
-     * @return FutureHttp
+     * @return Promise
      */
-    final protected function sendRequest(RequestInterface $requestInterface)
+    final protected function sendRequest(RequestInterface $request): Promise
     {
-        $runner = new FutureHttp($requestInterface);
-        $this->futureHttpPool->add($runner);
+        $test = Assertion::$currentTest;
+        $promise = \Amp\call(function () use($request) {
+            $req = new \Amp\Artax\Request($request->getUri(), $request->getMethod());
+            $req = $req->withProtocolVersions([$request->getProtocolVersion()]);
+            $req = $req->withHeaders($request->getHeaders());
+            $req = $req->withBody((string) $request->getBody());
+
+            /** @var Lock $lock */
+            $lock = yield $this->semaphore->acquire();
+            /** @var Response $response */
+            $response = yield $this->client->request($req);
+            $content = yield $response->getBody()->read();
+
+            $lock->release();
+
+            return $this->messageFactory->createResponse(
+                $response->getStatus(),
+                $response->getReason(),
+                $response->getHeaders(),
+                $content,
+                $response->getProtocolVersion()
+            );
+        });
+
+        $promise->onResolve(function () use ($test) {
+            Assertion::$currentTest = $test;
```

Comment:
If that's the only thing that uses the event loop, that's probably fine. But be careful if tests are allowed to do their own things with the event loop.

Reply:
Yes this is a temporary solution need to find a better way to attach assertion as metadata for a test

Reply:
If you want to use global state, you can use something like https://gist.github.com/brstgt/7f7e21ff7d7021f280c2932aebffab67

Reply:
Don't really need a global test, i just need to say that an assertion is linked to a specific test (or task / promise), but unfortunately i loose this information at the first async call

Reply:
You can also use instance methods or provide an assertion object to the test instead of using static methods for assertions. Or simply throw for all assertion errors, that should solve it, too.